cdc/scheduler/agent.go

```diff
@@ -0,0 +1,448 @@
+	}
+
+	if err := a.sendCheckpoint(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	if a.needSyncNow.Load() {
+		done, err := a.sendSync(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !done {
+			// We need to send a sync successfully before proceeding.
+			return nil
+		}
+		a.needSyncNow.Store(false)
+	}
+
+	opsToApply := a.popPendingOps()
+	for _, op := range opsToApply {
+		if _, ok := a.tableOperations[op.TableID]; ok {
+			a.logger.DPanic("duplicate operation", zap.Any("op", op))
+			return cerrors.ErrProcessorDuplicateOperations.GenWithStackByArgs(op.TableID)
+		}
+		a.tableOperations[op.TableID] = op
+	}
+
+	if err := a.processOperations(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// GetLastSentCheckpointTs implements the interface Agent.
+func (a *BaseAgent) GetLastSentCheckpointTs() model.Ts {
+	return a.checkpointSender.LastSentCheckpointTs()
+}
+
+func (a *BaseAgent) popPendingOps() (opsToApply []*agentOperation) {
+	a.pendingOpsMu.Lock()
+	defer a.pendingOpsMu.Unlock()
+
+	for !a.pendingOps.Empty() {
+		opsBatch := a.pendingOps.PopManyFront(128 /* batch size */)
+		for _, op := range opsBatch {
+			opsToApply = append(opsToApply, op.(*agentOperation))
+		}
+	}
+	return
+}
+
+// sendSync needs to be called with a.pendingOpsMu held.
+func (a *BaseAgent) sendSync(ctx context.Context) (bool, error) {
+	var adding, removing, running []model.TableID
+	for _, op := range a.tableOperations {
+		if !op.IsDelete {
+			adding = append(adding, op.TableID)
+		} else {
+			removing = append(removing, op.TableID)
+		}
+	}
+	for _, tableID := range a.executor.GetAllCurrentTables() {
+		if _, ok := a.tableOperations[tableID]; ok {
+			// Tables with a pending operation is not in the Running state.
+			continue
+		}
+		running = append(running, tableID)
+	}
+
+	// We are sorting these so that there content can be predictable in tests.
+	// TODO try to find a better way.
+	util.SorterTableIDs(running)
+	util.SorterTableIDs(adding)
+	util.SorterTableIDs(removing)
+	done, err := a.communicator.SyncTaskStatuses(ctx, running, adding, removing)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if !done {
+		return false, nil
+	}
+	return true, nil
+}
+
+// processOperations tries to make progress on each pending table operations.
+// It queries the executor for the current status of each table.
+func (a *BaseAgent) processOperations(ctx context.Context) error {
+	for tableID, op := range a.tableOperations {
+		switch op.status {
+		case operationReceived:
+			if !op.IsDelete {
+				// add table
+				if err := a.executor.AddTable(ctx, op.TableID); err != nil {
+					return errors.Trace(err)
+				}
+				op.status = operationProcessed
+			} else {
+				// delete table
+				done, err := a.executor.RemoveTable(ctx, op.TableID)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if done {
+					op.status = operationProcessed
+				} else {
+					break
+				}
+			}
+			fallthrough
```

Comment:
What about
- Remove L266
- Change L273-L277 to 
```go
if !done {
    break
}
```
- move `op.status = operationProcessed` before L279, then the state transition will be more clear.